doc/design/fully_static_graph.md

```diff
@@ -0,0 +1,122 @@
+# Design Doc: Fully Static Graph
```

Comment:
After discussing this design together with @reyoung, I noticed that this PR doesn't provide sufficient background information for readers to understand the problem.

As a design doc, this PR doesn't tell what is the problem. It directly goes to a conclusion. This is like firing a missile with no target.

It looks to me that this PR is trying to address the choice -- should we allow users to write the train loop, in hope that the PaddlePaddle program doesn't change much when running locally and distributedly.

To help readers understand this problem better, we can start from how a DL program looks like:

```python
reader = new_reader("some_files")
for iterations:
  minibatch = reader.next()
  train(a_network, minibatch)
```

If we allow users to write the train loop, the users should also write the creation of the reader because the loop uses it. This means that users specify the reading and feeding of data. However, it doesn't make sense all workers run this loop over the same data -- what we want is that each worker loads part of the data.

It is true that most DL frameworks make the reader depends on a 0-based successive "worker id", e.g., distributed PyTorch, and this approach works with HPC clusters. However, in a fault-tolerable distributed computing environment, there is no such a worker id.

This is the reason that we might need to restrict the user from being able to write the train loop.  And if we cannot make this restriction, how should we solve above problem.  It seems that this document presents a solution -- to wrap the train loop into a PaddlePaddle operator, other than a Python loop.

Reply:
Thanks for taking time to read the PR and write this very valuable feedback! It's very valuable to me. I will improve this PR to provide detailed background and reason behind the solution this PR proposes.

Reply:
Thanks again for the feedback! I provided detailed background and reason in the updated PR.